src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PathSeqFilterSpark.java

```diff
@@ -0,0 +1,192 @@
+
+        //Filter duplicate reads
+        final OpticalDuplicateFinder finder =
+                new OpticalDuplicateFinder(opticalDuplicatesArgumentCollection.READ_NAME_REGEX, opticalDuplicatesArgumentCollection.OPTICAL_DUPLICATE_PIXEL_DISTANCE, null);
+        JavaRDD<GATKRead> markedReads = MarkDuplicatesSpark.mark(primaryReads, getHeaderForReads(),MarkDuplicatesScoringStrategy.SUM_OF_BASE_QUALITIES, finder, getRecommendedNumReducers());
+
+        ReadFilter markedFilter = new MarkedDuplicateReadFilter();
+        final JavaRDD<GATKRead> markedFilteredReads = markedReads.filter(read -> markedFilter.test(read));
+
+        //Apply DUST masking
+        ReadTransformer dustTransformer = new DUSTReadTransformer();
+        final JavaRDD<GATKRead> readsDUSTMasked = markedFilteredReads.map(read -> dustTransformer.apply(read));
+
+        //Apply base quality clipping
+        ReadTransformer bqClippingTransformer = new BaseQualityClipReadTransformer();
+        final JavaRDD<GATKRead> readsClipped = readsDUSTMasked.map(read -> bqClippingTransformer.apply(read));
+
+        //Filter using base quality scores
+        ReadFilter bqFilter = new BaseQualityReadFilter(QUAL_PHRED_THRESH,QUAL_MAX_FRAC);
+        final JavaRDD<GATKRead> readsBQFiltered = readsClipped.filter(read -> bqFilter.test(read));
+
+        //Filter reads with ambiguous bases
+        ReadFilter ambigFilter = new AmbiguousBaseReadFilter(FRAC_N_THRESHOLD);
+        final JavaRDD<GATKRead> readsAmbigFiltered = readsBQFiltered.filter(read -> ambigFilter.test(read));
+
+        //Filter reads with less than MIN_READ_LENGTH bases
+        ReadFilter readLengthFilter = new ReadLengthReadFilter(MIN_READ_LENGTH,MAX_READ_LENGTH);
+        final JavaRDD<GATKRead> readsLengthFiltered = readsAmbigFiltered.filter(read -> readLengthFilter.test(read));
+
+        //Load Kmer hopscotch set and filter reads containing > 0 matching kmer
+        final PipelineOptions options = getAuthenticatedGCSOptions();
+        Input input = new Input(BucketUtils.openFile(KMER_LIB_PATH, options));
+        Kryo kryo=new Kryo();
+        HopscotchSet<SVKmerSmall> kmerLibSet = (HopscotchSet<SVKmerSmall>)kryo.readClassAndObject(input);
+        Broadcast<HopscotchSet<SVKmerSmall>> kmerLibBroadcast = ctx.broadcast(kmerLibSet);
+        ContainsKmerReadFilter readKmerFilter = new ContainsKmerReadFilter(kmerLibBroadcast,KMER_SIZE);
+        final JavaRDD<GATKRead> readsKmerFiltered = readsLengthFiltered.filter(read -> !readKmerFilter.test(read));
+
+        //TODO: bwa filtering against user-specified reference
+
+        //"Clean" reads by resetting all flags and setting pairedness flags according to the remaining read set
+        final JavaRDD<GATKRead> readsCleaned = readsKmerFiltered
+                .groupBy(read -> read.getName()) //group pairs using read name into set of Iterable<GATKRead>'s
```

Comment:
This could be really, really expensive for an input with a lot of bacterial reads.  Shuffling a large number of GATKReads is crazy expensive.  (But maybe you're confident that the filtered set will be small?)
I don't have code, or even a detailed design to try to circumvent this, but a sketch of an idea might go something like this:
Persist the readsKmerFiltered RDD.  Map it to just the read names, and shuffle them. Within each partition save just the unpaired names (or just the paired names -- whichever set you expect to be smaller).  Broadcast the set of unpaired (or paired) names.  Now pass over the same RDD you persisted earlier, looking up each name in the set to tell you how to "clean" it.